src/core/meshcop/tcat_agent.cpp

```
@@ -506,45 +514,106 @@ Error TcatAgent::HandleSingleTlv(const Message &aIncomingMessage, Message &aOutg
+    {
+        error = kErrorRejected;
+        ExitNow();
+    }
+
+    SuccessOrExit(error = Get<ActiveDatasetManager>().Read(datasetTlvs));
+    SuccessOrExit(
+        error = Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, datasetTlvs.mTlvs, datasetTlvs.mLength));
+    aResponse = true;
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandleGetCommissionerCertificate(Message &aOutgoingMessage, bool &aResponse)
+{
+    Error         error = kErrorNone;
+    Dataset       dataset;
+    unsigned char buf[kCommissionerCertMaxLength];
+    uint16_t      bufLen = sizeof(buf);
+
+    if (!CheckCommandClassAuthorizationFlags(mCommissionerAuthorizationField.mCommissioningFlags,
+                                             mDeviceAuthorizationField.mCommissioningFlags, &dataset))
+    {
+        error = kErrorRejected;
+        ExitNow();
+    }
+
+    VerifyOrExit(kErrorNone == Get<Settings>().GetTcatCommissionerCertificate(buf, bufLen), error = kErrorInvalidState);
+    SuccessOrExit(error = Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, buf, bufLen));
+    aResponse = true;
+
 exit:
     return error;
 }
 
 Error TcatAgent::HandleDecomission(void)
 {
-    Error error = kErrorNone;
+    Error         error = kErrorNone;
+    unsigned char buf[kCommissionerCertMaxLength];
+    size_t        bufLen = sizeof(buf);
+    Dataset       dataset;
+
+    if (!CheckCommandClassAuthorizationFlags(mCommissionerAuthorizationField.mDecommissioningFlags,
+                                             mDeviceAuthorizationField.mDecommissioningFlags, &dataset))
+    {
+        error = kErrorRejected;
+        ExitNow();
+    }
 
     IgnoreReturnValue(otThreadSetEnabled(&GetInstance(), false));
     Get<ActiveDatasetManager>().Clear();
     Get<PendingDatasetManager>().Clear();
 
     error = Get<Instance>().ErasePersistentInfo();
 
+    SuccessOrExit(error = Get<Ble::BleSecure>().GetPeerCertificateDer(buf, &bufLen, bufLen));
```

If this call fails, the below clearing of the network key (lines 540-546) still needs to be done. Otherwise, if there's a failure getting the Peer cert, then the decommission is "half" done leaving the device in an invalid state.  (I think)

---

A related issue is that when the `GetPeerCertificateDer` would fail, the decommission operation is mostly (/half) done while the returned status code error suggests that it did not succeed. Again, this may not be the right type of feedback to the TCAT-commissioner/user.

One easy solution is to ignore any error in getting the peer cert and declare the operation as a success in case the datasets/keys got wiped.  Again asking @arnulfrupp  for feedback.

---

Normally successful operation goes before successful logging. However, the purpose of storing the commissioner certificate is traceability. A commissioner should not be able to perform potentially harmful configuration changes and hide his identity, e.g. by using up flash memory using other commands before.

I suggest we should first make an attempt to store the certificate. Only if storing the certificate was successful, the decommissioning operations should be performend. Luckily the dataset and key delete operations cannot fail, so documentation and operation will always be in sync.

If the commissioner (repeatedly) receives an error on decommissioning, then the device should be factory reset (not in scope of this command), sent to the vendor for maintenance or scraped.

---

@arnulfrupp that's ok for me, agree. In this case the device effectively keeps track of which commissioner last *attempted* the operation :) , which is the most useful in any case.  

One practical case where storing the cert in flash may fail is if the TCAT commissioner's cert is slightly too large (>1024B).

---

A certificate >1024B is out of spec. Returning an error in that case seems appropriate.

---

@arnulfrupp  Yes, agree - instead of sending the TCAT device for maintenance, the commissioner needs to be sent back :)

---

Please check the new version.